File: src/detector/fasterRcnn/FasterRcnnDetectorDemo.m
clear all

%% set paths
assert (~isempty(getenv('CITY_DATA_PATH')));  % make sure environm. var set
CITY_DATA_PATH = [getenv('CITY_DATA_PATH') '/'];    % make a local copy
addpath(genpath(fullfile(getenv('CITY_PATH'), 'src')));  % add tree to search path
cd (fileparts(mfilename('fullpath')));        % change dir to this script



%% input
model_dir = fullfile(getenv('FASTERRCNN_ROOT'), 'output/faster_rcnn_final/faster_rcnn_VOC0712_ZF');
image_path  = '../testdata/image00001.png';


%% init

% detector
fasterRcnnDetector = FasterRcnnDetector (model_dir, 'use_gpu', true);

%% work 

img =  imread(image_path);

tic
%fasterRcnnDetector.setVerbose(1);
cars = fasterRcnnDetector.detect(img);
toc

cmap = colormap('Autumn');
for i = 1 : length(cars)
    colorindex = floor(cars(i).score * size(cmap,1)) + 1;
    color = cmap (colorindex, :) * 255;
    img = cars(i).drawCar(img, 'color', color);
end
imwrite(img, 'resultdemo.jpg');
%imshow([mask2rgb(mask), img]);

File: src/detector/fasterRcnn/FasterRcnnDetectorDemo.py
import logging
import sys, os, os.path as op
import numpy as np
import cv2
import json
from FasterRcnnDetector import FasterRcnnDetector
import matplotlib.pyplot as plt
from utils.timer import Timer
sys.path.insert(0, op.join(os.getenv('CITY_PATH'), 'src/car'))
from Car import Car



def vis_detections(im, cars):
    """Draw detected bounding boxes."""

    im = im[:, :, (2, 1, 0)]
    fig, ax = plt.subplots(figsize=(12, 12))
    ax.imshow(im, aspect='equal')
    for car in cars:
        bbox = car.bbox
        score = car.score
        name = car.name

        ax.add_patch(
            plt.Rectangle((bbox[0], bbox[1]), bbox[2], bbox[3],
                          fill=False,
                          edgecolor='red', linewidth=3.5)
            )
        ax.text(bbox[0], bbox[1] - 2,
                '{:s} {:.3f}'.format(name, score),
                bbox=dict(facecolor='blue', alpha=0.5),
                fontsize=14, color='white')

    plt.axis('off')
    plt.tight_layout()
    plt.draw()
    plt.show()


if __name__ == '__main__':
    ''' demo with a sample image'''

    # input
    model_dir = 'vgg16'
    image_path  = '../testdata/image00001.png'

    # init
    logging.basicConfig (level=logging.INFO)
    im = cv2.imread(image_path)
    detector = FasterRcnnDetector (model_dir, cpu_mode = True)
    detector.CONF_THRESH = 0.5


    # work
    timer = Timer()
    timer.tic()
    cars = detector.detect(im)
    timer.toc()
    print 'detection took %.3f sec' % timer.total_time

    print 'got %d detections' % len(cars)

    # visualize detections for each class
    for car in cars:
        print car.bbox, car.score, car.name
    vis_detections(im, cars)
